packages/cloudflare/src/api/overrides/queue/queue-cache.ts

```diff
@@ -0,0 +1,123 @@
+  readonly regionalCacheTtlSec: number;
+  readonly waitForQueueAck: boolean;
+  cache: Cache | undefined;
+  // Local mapping from key to insertedAtSec
+  localCache: Map<string, number> = new Map();
+
+  constructor(
+    private originalQueue: Queue,
+    options: QueueCachingOptions
+  ) {
+    this.name = `cached-${originalQueue.name}`;
+    this.regionalCacheTtlSec = options.regionalCacheTtlSec ?? DEFAULT_QUEUE_CACHE_TTL_SEC;
+    this.waitForQueueAck = options.waitForQueueAck ?? false;
+  }
+
+  async send(msg: QueueMessage) {
+    try {
+      const isCached = await this.isInCache(msg);
+      if (isCached) {
+        return;
+      }
+      if (!this.waitForQueueAck) {
+        await this.putToCache(msg);
+        await this.originalQueue.send(msg);
+      } else {
+        await this.originalQueue.send(msg);
+        await this.putToCache(msg);
+      }
+    } catch (e) {
+      error("Error sending message to queue", e);
+    } finally {
+      this.clearLocalCache();
+    }
+  }
+
+  private async getCache() {
+    if (!this.cache) {
+      this.cache = await caches.open("durable-queue");
+    }
+    return this.cache;
+  }
+
+  private getCacheUrlString(msg: QueueMessage) {
+    return `queue/${msg.MessageGroupId}/${msg.MessageDeduplicationId}`;
+  }
+
+  private getCacheKey(msg: QueueMessage) {
+    return "http://local.cache" + this.getCacheUrlString(msg);
+  }
+
+  private async putToCache(msg: QueueMessage) {
+    this.localCache.set(this.getCacheUrlString(msg), Date.now());
+    const cacheKey = this.getCacheKey(msg);
+    const cache = await this.getCache();
+    await cache.put(
+      cacheKey,
+      new Response(null, {
+        status: 200,
+        headers: {
+          "Cache-Control": `max-age=${this.regionalCacheTtlSec}`,
+          // Tag cache is set to the value of the soft tag assigned by Next.js
+          // This way you can invalidate this cache as well as any other regional cache
+          "Cache-Tag": `_N_T_/${msg.MessageBody.url}`,
+        },
+      })
+    );
+  }
+
+  private async isInCache(msg: QueueMessage) {
+    if (this.localCache.has(this.getCacheUrlString(msg))) {
+      const insertedAt = this.localCache.get(this.getCacheUrlString(msg))!;
+      if (Date.now() - insertedAt < this.regionalCacheTtlSec * 1000) {
+        return true;
+      }
+      this.localCache.delete(this.getCacheUrlString(msg));
+      return false;
+    }
+    const cacheKey = this.getCacheKey(msg);
+    const cache = await this.getCache();
+    const cachedResponse = await cache.match(cacheKey);
+    if (cachedResponse) {
+      return true;
+    }
+  }
+
+  /**
+   * Remove any value older than the TTL from the local cache
+   */
+  private clearLocalCache() {
+    const now = Date.now();
+    const insertAtSecMax = this.regionalCacheTtlSec * 1000;
+    for (const [key, insertAtSec] of this.localCache.entries()) {
+      if (now - insertAtSec > insertAtSecMax) {
+        this.localCache.delete(key);
+      }
```

Comment:
I meant something like

```suggestion
    const insertAtSecMax = Date.now() - this.regionalCacheTtlSec * 1000;
    for (const [key, insertAtSec] of this.localCache.entries()) {
      if (insertAtSec < insertAtSecMax) {
        this.localCache.delete(key);
      }
```

Reply:
Ho yeah right, sorry i forgot to turn on my brain :smile: